### ISSUE TITLE
[P1] Dense TP loading mistakes an undersized wire for a replicated layer when global shape is ignored

### ISSUE THREAD

**Issue by RobTand** [RobTand/tessera#303]
Source reviewed: `RobTand/tessera` at `4e9744f6424ecd0d90ebcbd4200c909123f38a77`, compared with `d52e1ac`. This is an audit-only finding; the user requested issues and closure verification, not fixes.

**Trigger and impact.** A ColumnParallel layer at TP=4 requires a global weight `[256,64]` and a local `[64,64]` tile, but the checkpoint declares an entire `[64,64]` wire. The dense methods receive `input_size=64`, `output_size=256` and the actual layer with `tp_size=4`, yet discard those facts and plan only from local dimensions. Since local dimensions equal the wire dimensions, `plan_shard` treats the wire as a replicated whole layer on every rank. All four ranks then use rows `[0,64)` instead of owning four distinct ranges of a 256-row weight. Per-role cardinality and boundary checks pass. This is a malformed-checkpoint acceptance leading to wrong global Linear outputs, not a demonstrated multi-rank regression in a correctly shaped checkpoint.

**Evidence.** `src/tessera/serving/fp8_route.py:248` accepts the full dimensions; its call at line 259 forwards neither them nor layer-local TP identity. `nvfp4_route.py:117`/130 and `bf16_route.py:693`/701 do the same. `serving/sharding.py:442` obtains process-global TP state, and the whole-shape shortcut at line 494 infers replication solely from local dimensions. The subsequent output-cut branch also permits any role to have fewer distinct shards than TP ranks based on divisibility alone (line 530), without the runtime's actual KV replication contract.

**CPU reproduction.** Run from the repository with `PYTHONDONTWRITEBYTECODE=1 PYTHONPATH=src OMP_NUM_THREADS=1 MKL_NUM_THREADS=1 OPENBLAS_NUM_THREADS=1 python3`. This uses minimal vLLM base-class substitutes to exercise the real Tessera construction method; it does not run vLLM, allocate a GPU, parse a fake payload, or claim a served result.

```python
import sys, types, torch
# Stand in only for vLLM's base classes; run Tessera's real builder and
# create_weights, with the explicit shapes a ColumnParallel layer supplies.
for name in ["vllm", "vllm.model_executor", "vllm.model_executor.layers",
             "vllm.model_executor.layers.linear", "vllm.model_executor.parameter"]:
    sys.modules[name] = types.ModuleType(name)
sys.modules["vllm.model_executor.layers.linear"].LinearMethodBase = type(
    "LinearMethodBase", (), {})
class Parameter(torch.nn.Parameter):
    def __new__(cls, data, weight_loader=None):
        return super().__new__(cls, data, requires_grad=False)
sys.modules["vllm.model_executor.parameter"].BasevLLMParameter = Parameter
from tessera.serving.fp8_route import build_tessera_fp8_method
from tessera.serving import sharding
scheme = dict(family="TESSERA_FP8", grid="E4M3", body="WINDOW",
              plane="CHANNEL", q256=1024, rows=64, columns=64,
              wire_bytes=1, roles=[["weight", 64]])
class ColumnParallelLinear(torch.nn.Module):
    def __init__(self, rank):
        super().__init__()
        self.tp_size, self.tp_rank = 4, rank
        self.input_size, self.output_size = 64, 256
for rank in range(4):
    sharding.tp_rank_and_size = lambda rank=rank: (rank, 4)
    layer = ColumnParallelLinear(rank)
    method = build_tessera_fp8_method(scheme, "linear", "resident")
    method.create_weights(
        layer, input_size_per_partition=64, output_partition_sizes=[64],
        input_size=64, output_size=256, params_dtype=torch.bfloat16)
    plan = layer.tessera_shard_plan
    print(rank, plan.axis, plan.roles[0].lo, plan.roles[0].hi)
```

Observed output: `0 None 0 64`, `1 None 0 64`, `2 None 0 64`, `3 None 0 64`. A required global 256-row layer is registered with a whole 64-row wire on every rank. The one-byte wire declaration is immaterial to this construction-time shape proof; a complete valid 64x64 wire agrees with the same declaration and still supplies no missing 192 rows.

**Relation to closed #234 / PR #262.** The original same-total/per-role-boundary defect is fixed. This is a pre-existing sibling omission left outside that fix: local per-role extents can agree perfectly while the global runtime geometry disagrees. No reopening is needed for the specific original reproducer. TP>1 remains unattested in the contract; this finding concerns the executable loader's fail-closed behavior, not an assertion that existing TP>1 receipts are wrong.

**Why existing tests miss it.** Tests distinguish correctly sized replicated layers from correctly sized sharded layers by feeding planner inputs consistent with the intended answer. They do not provide a real `create_weights` full-size argument that contradicts the wire while local dimensions match. The new #234 tests exercise per-role size and count differences, not loss of the global dimension/ownership contract.

**Fix and acceptance.** Derive replication and the expected complete role geometry from the actual layer's parallel contract, forwarding the full-size inputs instead of inferring everything from local dimensions. Refuse undersized whole wires for ordinary ColumnParallel/RowParallel layers before registering a loadable method. Preserve genuine `disable_tp`/replicated Linears and GQA/MQA KV head replication; a simple comparison against QKV's padded aggregate output_size is insufficient. Tests should show these malformed cases fail while valid TP1, replicated, even-split and GQA/MQA cases retain their correct rank ownership. No new serve measurement is needed to establish the named shape refusal; any broader TP serving claim still needs its own multi-rank evidence.

No source fixes were made because this session is expressly audit-only.

Coordinator delta check: fetched Tessera master b444631 during this review. Changes since the pinned audit snapshot in fused.py, trellis.py and merge_suite.py are comments/diagnostic prose; the reported behavior remains. #283 is now closed by PR284, which explicitly documents the divergence bound as policy without changing the min join.

Disposition: filed during the requested review of closed audit issues and current code. No implementation changes were made; fixes are left for a separate dispatch.
